text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,550 @@
+   version they have published, they might end up hitting this limit because of
+   the large size of `aws-cdk-lib`.
+6. The developer experience of using `aws-cdk-lib` in an IDE might be slower in
+   some cases.
+7. The CDK project is currently blocked from including more ‘large
+   dependencies’, such as multiple versions of kubectl, multiple versions of the
+   AWS CLI, and dedicated versions of the AWS SDK for use in custom resources.
+   This is a decision we’ve made to avoid increasing the size of the package
+   further until a more scalable solution for these kinds of dependencies is
+   implemented.
+
+Most of these above issues have workarounds. For running CDK apps in Lambda, you
+can use a Docker image for your Lambda code, and the limit becomes 10GB. For
+package manager size limits, we can request
+[increases](https://github.com/pypa/pypi-support/issues/1642), and those have
+historically been approved. However, publishing a large package is a bad
+practice in the open-source software ecosystem. Even if a customer has no
+technical limitations for adding a 234 MB dependency to their package, it is a
+red flag to them when considering using `aws-cdk-lib`. We should be
+customer-obsessed, and solve the problem of `aws-cdk-lib` being a huge
+dependency, before customers start complaining about it.
+
+### Why is the package so large today?
+
+There are some large assets contributing to the size of `aws-cdk-lib`, like
+.jsii files, and zip files of dependencies used in custom resources. The library
+also contains a huge volume of source code now that `aws-cdk-lib` combines 231
+(and increasing) CDK modules. The table below breaks down the percentage that
+each category of files contributes to the size of aws-cdk-lib.
+
+Note: The percentages do not add up to 100. Some files, e.g. various .json
+files, are excluded and contribute very little to the size.
+
+| Category                       | Percentage of aws-cdk-lib size |
+| ------------------------------ | ------------------------------ |
+| .jsii files                    | 41.64%                         |
+| Source Maps (.js.map)          | 21.93%                         |
+| Lambda Layer zip files         | 15.48%                         |
+| Javascript code (.js)          | 9.17%                          |
+| Type Declaration files (.d.ts) | 9.15%                          |
+| README, etc (.md)              | 0.76%                          |
+| bundled npm dependencies       | 0.69%                          |
+| .ts-fixture                    | 0.37%                          |
+| .jsiirc                        | 0.36%                          |
+
+### Why is the package going to be larger in the future?
+
+The CDK team is currently blocked on adding more Lambda Layer zip files to the
+AWS CDK framework, because adding them would dramatically increase the size of
+the package. Each of these zips are in a module whose only purpose is to bundle
```

Comment:
I'm not sure I fully understand the plan here: Is the plan to reduce the package size so that it can be increased again with these layers? If so, then the arguments here are contradictory: We won't achieve both adding the layers and the benefits mentioned above of smaller package size at the same time.
Also, then the problem will come up again once there is a new layer which reaches the limit again.

Reply:
I need to clarify this better in the doc, since you are asking, it must not be clear! But the goal is to remove large dependencies for these layers, and also create a mechanism for adding more layers in the same manner. So that we can add more Lambda Layer dependencies without ballooning the package size after this project is implemented.

Reply:
I've added some more detail in this section to explain. Thanks for your comment @dbartholomae!